File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline

N, M, V = map(int, input().split())

graph = [[0] * (N + 1) for _ in range(N + 1)]
for _ in range(M):
    a, b = map(int, input().split())
    graph[a][b] = 1
    graph[b][a] = 1

visited_DFS = [0] * (N + 1)
visited_BFS = [0] * (N + 1)


def dfs(start):
    visited_DFS[start] = 1
    print(start, end=" ")

    for i in range(1, N + 1):
        if graph[start][i] == 1 and visited_DFS[i] == 0:
            dfs(i)


def bfs(start):
    queue = deque([])
    queue.append(start)

    while queue:
        _start = queue.popleft()
        visited_BFS[_start] = 1
        print(_start, end=" ")

        for i in range(1, N + 1):
            if graph[_start][i] == 1 and visited_BFS[i] == 0:
                visited_BFS[i] = 1
                queue.append(i)


dfs(V)
print()
bfs(V)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLines = [];

rl.on("line", (line) => {
  inputLines.push(line);
});

rl.on("close", () => {
  const [N, M, V] = inputLines[0].split(" ").map(Number);
  const graph = new Graph(N);

  for (let i = 1; i <= M; i++) {
    const [a, b] = inputLines[i].split(" ").map(Number);
    graph.addEdge(a, b);
  }

  console.log(graph.dfs(V).join(" "));
  console.log(graph.bfs(V).join(" "));
});

class Graph {
  constructor(N) {
    this.N = N;
    this.adjacencyMatrix = Array.from(Array(N + 1), () => Array(N + 1).fill(0));
    this.visitedDFS = Array(N + 1).fill(false);
    this.visitedBFS = Array(N + 1).fill(false);
  }

  addEdge(i, j) {
    this.adjacencyMatrix[i][j] = 1;
    this.adjacencyMatrix[j][i] = 1;
  }

  dfs(V) {
    this.visitedDFS[V] = true;
    const result = [V];

    for (let i = 1; i <= this.N; i++) {
      if (this.adjacencyMatrix[V][i] === 1 && !this.visitedDFS[i]) {
        result.push(...this.dfs(i));
      }
    }
    return result;
  }

  bfs(V) {
    let queue = [V];
    this.visitedBFS[V] = true;
    const result = [];

    while (queue.length > 0) {
      const node = queue.shift();
      result.push(node);

      for (let i = 1; i <= this.N; i++) {
        if (this.adjacencyMatrix[node][i] === 1 && !this.visitedBFS[i]) {
          queue.push(i);
          this.visitedBFS[i] = true;
        }
      }
    }
    return result;
  }
}
